### PR TITLE
feat(sink): support sink executor in force-append-only mode

### PR DESCRIPTION
**Pull request by xx01cyx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Support sink executor in force-append-only mode.
- Introduce `SinkType` for better append-only indication.
- Use `SinkType` to construct the corresponding sink connector.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

## Related issue

#7414

### CODE REVIEW

File: src/connector/src/sink/catalog/mod.rs [hunk 1]
```diff
@@ -54,6 +54,54 @@ impl From<SinkId> for u32 {
...
+    /// allowed.
+    AppendOnly,
+    /// The input of the sink operator can be INSERT, UPDATE, or DELETE, but it must drop any
+    /// UPDATE or DELETE and write only INSERT into the sink connector.
+    ForceAppendOnly,
+    /// The data written into the sink connector can be INSERT, UPDATE, or DELETE.
+    Upsert,
+    /// The sink can be either append-only or upsert. We should try append-only first. If
+    /// unavailable, fall back to upsert.
+    Any,
```
Comment (tabVersion) [inline src/connector/src/sink/catalog/mod.rs:69]: in which case will SinkType be any

**Review by tabVersion — APPROVED** [review]
rest LGTM

File: src/connector/src/sink/catalog/mod.rs (hunk 1 above)
Comment (xx01cyx) [reply src/connector/src/sink/catalog/mod.rs:69]: When the frontend derivation is append-only and the user also defines the sink as append-only. In this case, the sink **should** be append-only, but it's also correct that it is upsert.

File: src/frontend/src/optimizer/plan_node/stream_materialize.rs [hunk 2]
```diff
@@ -281,7 +282,11 @@ impl StreamMaterialize {
 
     /// Rewrite this plan node into [`StreamSink`] with the given `properties`.
     pub fn rewrite_into_sink(self, properties: WithOptions) -> StreamSink {
-        StreamSink::new(self.input, self.table.to_sink_desc(properties))
+        // TODO(Yuanxin): Deduce sink type here.
+        StreamSink::new(
+            self.input,
+            self.table.to_sink_desc(properties, SinkType::Any),
+        )
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/stream_materialize.rs:289]: IIUC, it is just an intermediate state of the implementation, but can we remove the `SinkType::Any`? I do not think we need it anywhere

File: proto/catalog.proto [hunk 3]
```diff
@@ -67,10 +75,10 @@ message Sink {
...
   // pk_indices of the corresponding materialize operator's output.
   repeated int32 stream_key = 9;
-  bool append_only = 10;
-  uint32 owner = 11;
-  map<string, string> properties = 12;
-  string definition = 13;
+  uint32 owner = 10;
+  map<string, string> properties = 11;
+  string definition = 12;
+  SinkType sink_type = 13;
```
Comment (st1page) [inline proto/catalog.proto:81]: I think we need not change too many fields here and the field number does not need to be compacted and ordered in fact....

File: src/frontend/src/optimizer/plan_node/stream_materialize.rs (hunk 2 above)
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/stream_materialize.rs:289]: will we support a connector that can accept both append-only stream and upsert stream?

Comment (st1page) [reply src/frontend/src/optimizer/plan_node/stream_materialize.rs:289]: Oh I have seen the comment, I think we can keep it.

Comment (xx01cyx) [reply src/frontend/src/optimizer/plan_node/stream_materialize.rs:289]: > IIUC, it is just an intermediate state of the implementation
Yes. This should be refactored later. But we have to deduce sink type somewhere in our code.

**Pull request merged**